File: array/183.cpp
// https://leetcode.com/problems/fair-candy-swap/
#include <bits/stdc++.h>

using namespace std;

using vi = vector<int>;
vector<int> fairCandySwap(vi& aliceSizes, vi& bobSizes) {
    int amountAlice = accumulate(aliceSizes.begin(), aliceSizes.end(), 0);
    int amountBob = accumulate(bobSizes.begin(), bobSizes.end(), 0);
    int half = (amountAlice + amountBob) / 2;
    int diff = abs(amountAlice - half);

    for(int i = 0; i < aliceSizes.size(); ++i) {
        for(int j = 0; j < bobSizes.size(); ++j) {
            if(amountAlice < half) {
                if(diff + aliceSizes[i] == bobSizes[j])
                    return {aliceSizes[i], bobSizes[j]};
            } else {
                if(diff + bobSizes[j] == aliceSizes[i])
                    return {aliceSizes[i], bobSizes[j]};
            }
        }
    }


    return {0, 0};
}

int main() {
    vector<int> alice = {1,3} ;
    vector<int> bob = {2};

    auto ans = fairCandySwap(alice, bob);
    cout<<ans[0]<<" "<<ans[1];

    return 0;
}

File: array/183.py
from typing import List


# O(n + m)
def fairCandySwap(aliceSizes: List[int], bobSizes: List[int]) -> List[int]:
    a = set(aliceSizes)
    b = set(bobSizes)
    alice_sum = sum(aliceSizes)
    bob_sum = sum(bobSizes)

    half = (alice_sum + bob_sum) // 2
    diff = abs(alice_sum - half)

    if alice_sum < half:
        for elem in aliceSizes:
            if elem + diff in b:
                return [elem, elem + diff]
    else:
        for elem in bobSizes:
            if elem + diff in a:
                return [elem + diff, elem]


print(fairCandySwap([2], [1, 3]))
